lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+                removalError = true;
+                return false;
+              });
+        })
+        .Case<FModuleOp>([&](FModuleOp op) {
+          // Handle annotations on the ports.
+          AnnotationSet::removePortAnnotations(
+              op, [&](unsigned i, Annotation annotation) {
+                if (!annotation.isClass(
+                        "sifive.enterprise.grandcentral.AugmentedGroundType"))
+                  return false;
+                auto maybeID = getID(op, annotation);
+                if (!maybeID)
+                  return false;
+                leafMap.insert({maybeID.getValue(), op.getArgument(i)});
+
+                return true;
+              });
+
+          // Handle annotations on the module.
+          AnnotationSet::removeAnnotations(op, [&](Annotation annotation) {
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.ViewAnnotation"))
+              return false;
+            auto tpe = annotation.getMember<StringAttr>("type"),
+                 name = annotation.getMember<StringAttr>("name");
+            auto id = annotation.getMember<IntegerAttr>("id");
+            if (!tpe) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain a 'type' field with a 'StringAttr' value";
+              goto FModuleOp_error;
+            }
+            if (!id) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain an 'id' field with an 'IntegerAttr' value";
+              goto FModuleOp_error;
+            }
+            if (!name) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain a 'name' field with a 'StringAttr' value";
+              goto FModuleOp_error;
+            }
+
+            // If this is a companion, then:
+            //   1. Insert it into the companion map
+            //   2. Create a new mapping module.
+            //   3. Instatiate the mapping module in the companion.
+            //   4. Check that the companion is instantated exactly once.
+            //   5. Set attributes on that lone instance so it will become a
+            //      bind if extraction information was provided.
+            if (tpe.getValue() == "companion") {
+              builder.setInsertionPointToEnd(circuitOp.getBody());
+
+              // Create the mapping module.
+              auto mappingName =
+                  getNamespace().newName(name.getValue() + "_mapping");
+              auto mapping = builder.create<FModuleOp>(
+                  circuitOp.getLoc(), builder.getStringAttr(mappingName),
+                  SmallVector<ModulePortInfo>({}));
```

Comment:
Is it possible to just pass in `{}`? Or we can try `ArrayRef<ModulePortInfo>()`.  I think constructing an empty small vector takes up some stack space. There are a bunch of these in a couple places.

Reply:
The latter works! Good thinking.